hcipy/optics/fiber.py

```diff
@@ -1,13 +1,144 @@
-	r2 = grid.x**2 + grid.y**2
-	return np.exp(-r2/((0.5 * mode_field_diameter)**2))
+class StepIndexFiber(AgnosticOpticalElement):
+	''' A step-index fiber.
+
+	The modal behaviour of the step-index fiber depends on the input parameters.
+
+	Parameters
+	----------
+	core_radius : scalar
+		The radius of the high-index core.
+	NA : scalar
+		The numerical aperture of the fiber.
+	fiber_length : scalar
+		The length of the optical fiber.
+	'''
+	def __init__(self, core_radius, NA, fiber_length):
+		self._core_radius = core_radius
+		self._NA = NA
+		self.fiber_length = fiber_length
+		
+		AgnosticOpticalElement.__init__(self, False, True)
+	
+	def make_instance(self, instance_data, input_grid, output_grid, wavelength):
+		monochromatic_V = self.V(wavelength)
+		instance_data.fiber_modes, instance_data.beta = make_LP_modes(input_grid, monochromatic_V, self.core_radius, mode_cutoff=None)
+		instance_data.Minv = np.linalg.pinv(instance_data.fiber_modes.transformation_matrix)
+
+	def num_modes(self, wavelength):
+		'''The approximate amount of modes of the fiber.
+		'''
+		V = self.V(wavelength)
+		return V**2 / 2
+
+	def V(self, wavelength):
+		'''The normalized frequency parameter for step-index fibers.
+		'''
+		return 2 * np.pi / wavelength * self.core_radius * self.NA
+
+	def mode_field_radius(self, wavelength):
+		'''The mode field radius of the fiber.
+		'''
+		V = self.V(wavelength)
+		w = self.core_radius * (0.65 + 1.619 / V**(3 / 2) + 2.879 / V**6)
+		return w
+
+	@property
+	def core_radius(self):
+		'''The core radius of this fiber.
+		'''
+		return self._core_radius
+	
+	@core_radius.setter
+	def core_radius(self, core_radius):
+		self._core_radius = core_radius
+		self.clear_cache()
+	
+	@property
+	def NA(self):
+		'''The numerical aperture of this fiber.
+		'''
+		return self._NA
+
+	@NA.setter
+	def NA(self, NA):
+		self._NA = NA
+		self.clear_cache()
+
+	def get_input_grid(self, output_grid, wavelength):
+		return output_grid
+
+	def get_output_grid(self, input_grid, wavelength):
+		return input_grid
+
+	@make_agnostic_forward
+	def modal_decomposition(self, instance_data, wavefront):
+		'''Decompose the input wavefront into the modal distribution of the fiber.
+
+		Parameters
+		----------
+		wavefront : Wavefront
+			The incoming wavefront.
+		
+		Returns
+		-------
+		array_like
+			The modal coefficients.
+		'''
+		wf = wavefront.copy()
+		
+		M = instance_data.fiber_modes.transformation_matrix
+		mode_coefficients = M.dot(wf.electric_field * wf.grid.weights)
+		return mode_coefficients
+
+	@make_agnostic_forward
+	def forward(self, instance_data, wavefront):
+		'''Forward propagate the light through the fiber.
+
+		Parameters
+		----------
+		wavefront : Wavefront
+			The incoming wavefront.
+		
+		Returns
+		-------
+		Wavefront
+			The wavefront that exits the fiber.
+		'''
+		wf = wavefront.copy()
+		
+		M = instance_data.fiber_modes.transformation_matrix
+		mode_coefficients = M.dot(wf.electric_field.conj() * wf.grid.weights)
```

Comment:
Yes I need to check this. I was busy with this code when you guys merged the polarization upgrade into the master branch. I will modify it so that it will work with polarization.

Reply:
This should be resolved now. I changed it so that the fiber is agnostic about the tensor order of the field and works only on the spatial axis with einsum.

Reply:
It works for polarization now. I made the fiber agnostic to the tensor order of the field with the use of einsum.